sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -84,26 +91,102 @@ static RT::PiProgram createSpirvProgram(const RT::PiContext Context,
+  return WithBuildState->Ptr.load();
+}
+
 RT::PiProgram ProgramManager::getBuiltOpenCLProgram(OSModuleHandle M,
                                                     const context &Context) {
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
-  std::map<OSModuleHandle, RT::PiProgram> &CachedPrograms =
-      Ctx->getCachedPrograms();
-  auto It = CachedPrograms.find(M);
-  if (It != CachedPrograms.end())
-    return It->second;
 
+  using PiProgramT = kernel_program_cache::PiProgramT;
+  using ProgramWithBuildState = kernel_program_cache::ProgramWithBuildStateT;
+
+  kernel_program_cache &Cache = *Ctx->getKernelProgramCache().get();
+
+  ProgramWithBuildState *ProgramWithState;
+  bool InsertionTookPlace;
+
+  // put atomic pointer in program cache as data with nullptr before
+  // the program is actually built
+  {
+    auto Locked = Cache.acquireCachedPrograms();
+    auto &CachedPrograms = Locked.get();
+
+    auto Inserted = CachedPrograms.emplace(std::piecewise_construct,
+                                           std::forward_as_tuple(M),
+                                           std::forward_as_tuple(nullptr,
+                                                                 BS_None));
+
+    InsertionTookPlace = Inserted.second;
+    ProgramWithState = &Inserted.first->second;
+  }
+
+  // any thread which will find nullptr in cache will wait until the pointer
+  // is not null anymore
+  if (!InsertionTookPlace) {
+    return waitUntilBuilt(Cache, ProgramWithState);
+  }
+
+  {
+    int Expected = BS_None;
+    int Desired = BS_Building;
+
+    if (!ProgramWithState->State.compare_exchange_strong(Expected, Desired)) {
+      // Someone is already building or have built the program.
+      // We've got a funny story here
+      assert(false && "Invalid state of program build in programs cache");
+    }
+  }
+
+  // we may now build the program whilst other threads are waiting
   DeviceImage *Img = nullptr;
-  using PiProgramT = remove_pointer_t<RT::PiProgram>;
   unique_ptr_class<PiProgramT, decltype(&::piProgramRelease)> ProgramManaged(
-      loadProgram(M, Context, &Img),
-      RT::PluginInformation.PiFunctionTable.piProgramRelease);
+        nullptr, RT::PluginInformation.PiFunctionTable.piProgramRelease);
 
-  build(ProgramManaged.get(), Img->BuildOptions);
-  RT::PiProgram Program = ProgramManaged.release();
-  CachedPrograms[M] = Program;
+  try {
+    ProgramManaged = {loadProgram(M, Context, &Img),
+      RT::PluginInformation.PiFunctionTable.piProgramRelease};
 
-  return Program;
+    build(ProgramManaged.get(), Img->BuildOptions);
+
+    PiProgramT *Desired = ProgramManaged.release();
+    PiProgramT *Expected = nullptr;
+
+    if (!ProgramWithState->Ptr.compare_exchange_strong(Expected, Desired)) {
+      // We've got a funny story here
+      assert(false && "We've build a program that is already have been built.");
+    }
+
+    ProgramWithState->State.store(BS_Built);
```

Comment:
Can we use more relaxed memory models for all loads and stores you are adding?

Reply:
I don't think we may use relaxed memory model here as current (sequentially consistent model) makes other threads to synchronize on this variable.